pkg/controllers/configobservation/customroutersecret/observe_custom_router_secret.go

```diff
@@ -0,0 +1,135 @@
+	return (&customRouterObserver{
+		configPaths: [][]string{{"servingInfo", "componentRoutes"}},
+		destSecret:  "v4-0-config-system-custom-router-certs",
+		componentRoute: types.NamespacedName{
+			Namespace: "openshift-authentication",
+			Name:      "oauth",
+		},
+		resourceType: corev1.Secret{},
+	}).observe
+}
+
+type customRouterObserver struct {
+	configPaths    [][]string
+	destSecret     string
+	componentRoute types.NamespacedName
+	resourceType   interface{}
+}
+
+// extractPreviouslyObservedConfig extracts the previously observed config from the existing config.
+func extractPreviouslyObservedConfig(existing map[string]interface{}, paths ...[]string) (map[string]interface{}, []error) {
+	var errs []error
+	previous := map[string]interface{}{}
+	for _, fields := range paths {
+		value, found, err := unstructured.NestedFieldCopy(existing, fields...)
+		if !found {
+			continue
+		}
+		if err != nil {
+			errs = append(errs, err)
+		}
+		err = unstructured.SetNestedField(previous, value, fields...)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return previous, errs
+}
+
+func (c *customRouterObserver) observe(genericlisters configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (ret map[string]interface{}, _ []error) {
+	listers := genericlisters.(configobservation.Listers)
+	errs := []error{}
+
+	// pick the correct resource sync function
+	resourceSync := listers.ResourceSyncer().SyncSecret
+	if _, ok := c.resourceType.(corev1.ConfigMap); ok {
+		resourceSync = listers.ResourceSyncer().SyncConfigMap
+	}
+
+	previouslyObservedConfig, errs := extractPreviouslyObservedConfig(existingConfig, c.configPaths...)
+
+	ingress, err := listers.IngressLister.Get("cluster")
+	// if something went wrong, keep the previously observed config and resources
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedComponentRoutes, observedResources, err := c.componentRouteSecret(listers, ingress)
+	if err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	observedConfig := map[string]interface{}{}
+	if err := unstructured.SetNestedField(observedConfig, observedComponentRoutes, c.configPaths[0]...); err != nil {
+		return previouslyObservedConfig, append(errs, err)
+	}
+
+	errs = append(errs, syncObservedResources(resourceSync, observedResources)...)
+	return observedConfig, errs
+}
+
+// resourceSyncFunc syncs a resource from the source location to the destination location.
+type resourceSyncFunc func(destination, source resourcesynccontroller.ResourceLocation) error
+
+// syncActionRules rules define source resource names indexed by destination resource names.
+// Empty value means to delete the destination.
+type syncActionRules map[string]string
+
+// syncObservedResources copies or deletes resources, sources in GlobalUserSpecifiedConfigNamespace and destinations in OperatorNamespace namespace.
```

Comment:
this is the only mention about `GlobalUserSpecifiedConfigNamespace` in this repo

Reply:
Left over from a copy/paste, I'll rework the comments.